### PR TITLE
Add Sentry integration for Backend and Frontend

### PR DESCRIPTION
**Pull request by richard015ar**
Related issue: https://github.com/pressbooks/private/issues/374

This change implements a new Sentry Integration to PB, and also add a JavaScript observer to monitor performance and catch front end issues.
Details about how it works and documentation: https://docs.google.com/document/d/1hg35gRWlxcQqC2DElrBi2btLFQQXvMgs-NhgvmnMqpA/edit

### How to test
- Create and set the Environment variables following [docs](https://docs.google.com/document/d/1hg35gRWlxcQqC2DElrBi2btLFQQXvMgs-NhgvmnMqpA/edit) in your local environment (make sure `ENABLE_SENTRY=1`
- Force any error or exception in any enable plugin or theme within your code
- Check [Sentry pressbooks project](https://sentry.io/organizations/pressbooks/issues) issues and make sure the issue is there.

_Note there are new ENV variables for sentry and also we can  delete the previous one in Staging after apply this change._

### CODE REVIEW

**Comment by richard015ar** [comment]
A change was added: `twentyseventeen`  by `twentytwenty`  in test_allowedRootThemes. Looks like twentyseventeen isn’t found as is being filtered out of the input array, seems twentyseventeen is not bundled with newer WordPress versions.
Related issue: https://github.com/pressbooks/pressbooks/issues/2064

I will add more unit tests to cover >= 0.02% required.

File: assets/src/scripts/sentry.js [hunk 1]
```diff
@@ -0,0 +1,18 @@
+import * as Sentry from '@sentry/browser';
+import { Integrations } from '@sentry/tracing';
+
+/* global SentryParams */
+Sentry.init( {
+	dsn: SentryParams.dsn,
+	integrations: [ new Integrations.BrowserTracing() ],
+	tracesSampleRate: 1.0,
```
Comment (ho-man-chan) [inline assets/src/scripts/sentry.js:8]: curious how the sampling rate will affect server performance. Will observe during testing in staging server.

File: hooks.php [hunk 2]
```diff
@@ -346,11 +346,8 @@
 // Sentry initializer - Only for staging and production environments
 // -------------------------------------------------------------------------------------------------------------------
 if (
-	! is_null( env( 'WP_ENV' ) ) &&
-	WP_ENV !== 'development' &&
-	! is_null( env( 'SENTRY_KEY' ) ) &&
-	! is_null( env( 'SENTRY_ORGANIZATION' ) ) &&
-	! is_null( env( 'SENTRY_PROJECT' ) )
+	! is_null( env( 'ENABLE_SENTRY' ) ) &&
```
Comment (ho-man-chan) [inline hooks.php:349]: Concern for backwards compatibility since environment variables are removed and new ones are added.

File: inc/class-sentry.php [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+	}
+
+	/**
+	 * Get current user for tracking (if available)
+	 *
+	 * @return false|\WP_User
+	 */
+	public function getCurrentUserForTracking() {
+		$user_id = get_current_user_id();
+		$this->user = $user_id > 0 ? $user = get_userdata( $user_id ) : false;
```
Comment (ho-man-chan) [inline inc/class-sentry.php:74]: may reduce to 
`$this->user = $user_id > 0 ? get_userdata( $user_id ) : false;`

File: inc/class-sentry.php [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+				'environment' => env( 'WP_ENV' ) ?: self::DEFAULT_ENVIRNOMENT,
+				'user' => false,
+			];
+			if ( $this->user ) {
+				$script_params['user'] = [
+					'username' => $this->user->user_login,
+					'email' => $this->user->user_email,
+				];
+			}
+			wp_localize_script( self::WP_SCRIPT_NAME, 'SentryParams', $script_params );
```
Comment (ho-man-chan) [inline inc/class-sentry.php:151]: What's the purpose of this?

File: inc/utility/namespace.php [hunk 5]
```diff
@@ -1500,22 +1500,3 @@ function do_shortcode_by_tags( $content, array $tags ) {
 	$shortcode_tags = $_tags;
 	return $shortcoded;
 }
-
-function initialize_sentry() {
```
Comment (ho-man-chan) [inline inc/utility/namespace.php:1504]: Is this possible to reuse also the existing sentry environment variables so it's backward compatible?

File: tests/test-sentry.php [hunk 6]
```diff
@@ -1,22 +1,68 @@
...
+	public function setUp() {
+		parent::setUp();
+		$this->user = $this->factory()->user->create_and_get( [ 'role' => 'contributor' ] );
+		$this->sentry = new Sentry();
+	}
+
+	/**
+	 * Test init Sentry function to test instance returned.
+	 */
+	public function test_getInstance() {
```
Comment (ho-man-chan) [inline tests/test-sentry.php:33]: Additional negative test cases would increase the robustness of the feature.
Example:
1. Missing environment variable -> application still functioning as expected
2. No user signed in -> application still logs with user as null

Test to ensure the errors are actually tracked by Sentry would be a great validation of the feature working with simulated errors.

**Review by ho-man-chan — COMMENTED** [review]
Great use of OOP Concept! 👍 
I notice some concerns for backwards compatibility on the use of environment variable. It would be perfect if the previous environment variable can work with new features. This will greatly reduce the headache of Open source users.

Furthermore, I suggest to add testing negative cases (ex: missing environment variables or non logged user) to make sure the feature would still work and not explode.

Finally, a test to ensure simulated errors are actually sent to sentry would be perfect if possible...

File: assets/src/scripts/sentry.js (hunk 1 above)
Comment (richard015ar) [reply assets/src/scripts/sentry.js:8]: I suggest do it dynamically, so it can be configured different for each server. An new ENV variable could solve it, since it will depend on server hardware, bandwidth and other resources.

File: hooks.php (hunk 2 above)
Comment (richard015ar) [reply hooks.php:349]: It is compatible, since other variables (the previous ones) are not longer used, then are not considered in any part of the code.

File: inc/class-sentry.php (hunk 3 above)
Comment (richard015ar) [reply inc/class-sentry.php:74]: Yes! I will change it. It was an typing error, doesn't make sense how I did it, probably a typing error. Thank you for catching it!

File: inc/class-sentry.php (hunk 4 above)
Comment (richard015ar) [reply inc/class-sentry.php:151]: It is used for pass variables to the JavaScript script from server side before create and compile the script. More references here:
- https://wpshout.com/building-a-magical-golden-bridge-from-php-to-javascript-with-wp_localize_script/
- https://medium.com/vunamhung/passing-php-variables-to-javascript-with-wp-localize-script-d99e165c1f9c

File: inc/utility/namespace.php (hunk 5 above)
Comment (richard015ar) [reply inc/utility/namespace.php:1504]: yes, is possible. I can add that compatibility if it is convenient for you. Probably I would need document the reasons for future references. Could you help me with that?

File: tests/test-sentry.php (hunk 6 above)
Comment (richard015ar) [reply tests/test-sentry.php:33]: Makes sense! I will add those.

**Comment by ho-man-chan** [comment]
Appreciated. 🥇

File: assets/src/scripts/sentry.js [hunk 7]
```diff
@@ -0,0 +1,18 @@
+import * as Sentry from '@sentry/browser';
+import { Integrations } from '@sentry/tracing';
+
+/* global SentryParams */
+Sentry.init( {
+	dsn: SentryParams.dsn,
+	integrations: [ new Integrations.BrowserTracing() ],
+	tracesSampleRate: SentryParams.sample,
```
Comment (richard015ar) [inline assets/src/scripts/sentry.js:8]: Sample rate variable

File: inc/class-pressbookssentry.php [hunk 8]
```diff
@@ -0,0 +1,224 @@
...
+namespace Pressbooks;
+
+use function \Pressbooks\Utility\debug_error_log;
+use PressbooksMix\Assets;
+
+class PressbooksSentry {
+
+	const DEFAULT_ENVIRNOMENT = 'staging';
+
+	const DEFAULT_TRACE_SAMPLE_RATE = 0.5;
```
Comment (richard015ar) [inline inc/class-pressbookssentry.php:16]: Default value for sample rate

File: inc/class-pressbookssentry.php [hunk 9]
```diff
@@ -0,0 +1,224 @@
...
+
+	const WP_SCRIPT_NAME = 'script-sentry';
+
+	const ENABLED_VALUE = 1;
+
+	const ENVIRONMENT_VARIABLES_V1 = 1;
```
Comment (richard015ar) [inline inc/class-pressbookssentry.php:22]: See https://docs.google.com/document/d/1hg35gRWlxcQqC2DElrBi2btLFQQXvMgs-NhgvmnMqpA/edit#

File: inc/class-pressbookssentry.php [hunk 10]
```diff
@@ -0,0 +1,224 @@
...
+	 *
+	 * @return PressbooksSentry
+	 */
+	static public function init() {
+		if ( is_null( self::$instance ) ) {
+			self::$instance = new self();
+			self::$instance->setSentryDSNFromEnvironmentVariables();
+			self::$instance->getCurrentUserForTracking();
+			if (
+				self::environmentVariablesType() === self::ENVIRONMENT_VARIABLES_V1 ||
```
Comment (richard015ar) [inline inc/class-pressbookssentry.php:54]: If [previous integration version](https://github.com/pressbooks/private/issues/130#issuecomment-634938654) 
```
SENTRY_KEY=
SENTRY_ORGANIZATION=
SENTRY_PROJECT=
```
is implemented, only PHP issues will be captured by Sentry.

File: inc/class-pressbookssentry.php [hunk 11]
```diff
@@ -0,0 +1,224 @@
...
+	/**
+	 * Get type of environment variables defined.
+	 * V1 = SENTRY_KEY, SENTRY_ORGANIZATION and SENTRY_PROJECT.
+	 * V2 = ENABLE_SENTRY and SENTRY_DSN.
+	 *
+	 * @return false|int
+	 */
+	static public function environmentVariablesType() {
+		if ( self::areEnvironmentVariablesPresent() ) {
+			return (
```
Comment (richard015ar) [inline inc/class-pressbookssentry.php:101]: See: https://github.com/pressbooks/private/issues/130#issuecomment-634938654

File: tests/test-pressbooks.php [hunk 12]
```diff
@@ -33,10 +33,10 @@ public function test_allowedBookThemes() {
 	 * @group plugin
 	 */
 	public function test_allowedRootThemes() {
-		$result = $this->pb->allowedRootThemes( [ 'pressbooks-book' => true, 'pressbooks-clarke' => true, 'pressbooks-fake' => true, 'twentyseventeen' => true ] );
+		$result = $this->pb->allowedRootThemes( [ 'pressbooks-book' => true, 'pressbooks-clarke' => true, 'pressbooks-fake' => true, 'twentytwenty' => true ] );
```
Comment (richard015ar) [inline tests/test-pressbooks.php:36]: Fix for: https://github.com/pressbooks/pressbooks/issues/2064

**Comment by richard015ar** [comment]
@ho-man-chan, I implemented some changes:
- I added environment variables compatibility with [the previous ones](https://github.com/pressbooks/private/issues/130#issuecomment-634938654)
- I added more use cases considering negative cases and with different environment variables cases.
- I added `SENTRY_TRACE_SAMPLE_RATE` to be able to manipulate the sample rate using env variable. By default the value is 0.5.
- I updated the documentation: https://docs.google.com/document/d/1hg35gRWlxcQqC2DElrBi2btLFQQXvMgs-NhgvmnMqpA/edit#

**Review by ho-man-chan — APPROVED** [review]
Looks good!! Nice tests!
